### ISSUE TITLE
Bug with deletion.

### ISSUE THREAD

**Issue by malisper** [hargettp/hh-redblack#3]
I was trying out your library and discovered a bug with deletion. If you insert the elements -1, -7, 1, -1, and -4 and then delete the elements 1, -1, -7, you will enter an infinite loop. This is happening because the `parent` slot of the leaf node isn't handled properly by `rb-transplant`. You mention that in a comment that the persistent case doesn't work properly. If you give me a day or so, I should be able to piece together a pull request. There are also some other miscellaneous things I may fix in the pull request.

**Comment by hargettp** [comment]
Oh wow thank you for your interest!

I am *more* than happy to entertain a pull request if one were to arrive. 

Let me know if you need any help with the archaeology of what I was thinking. It's been a while and I was learning Lisp at the time, so disasters abound. 

:)

**Comment by malisper** [comment]
A couple things. What was the problem with the leaf node and persistent trees? Why do you have the user extend `rb-key-compare` and return a keyword determining the result instead of having them just extend `rb<` and `rb=`. Lastly, you mention in a comment in `rb-put` that `nil` cannot be stored. Why not? `rb-get` returns multiple values so the user could just test the second one to determine whether nil was actually stored.

**Comment by hargettp** [comment]
Re ```rb-key-compare```: good point, I have no idea why I thought that important. The only thing I can think of is that I was a fan of having a single function define comparison; but I'm not feeling religious about it. :)

Re ```rb-put```: the best recollection I have (and I'm not saying it matches the code) is that I used nil as the sentinel value in the red-black tree. Would not surprise me if I implemented it in such a way that that turned out to be unnecessary.

Will try to give the code a closer reading over the next few days, as I have forgotten so much of what is in here. :)

**Comment by malisper** [comment]
I have created a [fork](https://github.com/malisper/hh-redblack) in which I have fixed everything I have found wrong with the in memory implementation except for the use of `with-slots`. The problem with using `with-slots` is that it ignores any around methods defined on the accessor methods. I am now going to go through the implementation of persistent trees and try to clean that up.

**Comment by hargettp** [comment]
V cool, ty! Will take a look.

**Comment by malisper** [comment]
So I tried cleaning up the code for the persistent implementation. You do a lot of weird stuff since with `with-rb-transaction` transaction needs to exist when a node is created and a node is created when the tree is initialized. I found that it may be possible to use an around method on initialize-instance that wraps the initialization with `with-rb-transaction` which winds up being much simpler. This works except there are some issues with the text implementation because it writes to the file when the file is first opened. I am going to spend some time seeing if it is possible to re-architect the code. There also seems to be a lot of redundancy as you have `prb-load-node`, `prb-load-data`, `prb-fetch-node`, `prb-fetch-data`, etc.

**Comment by hargettp** [comment]
yeah, this was my first attempt at persistent data structures. fairly messy, as you can see. :)

my goal was to get the basic red-black tree algorithm correct, then just use the various tricks of Common Lisp to use the same algorithms when manipulating a persistent version. would not be surprised if I did not succeed.

**Comment by malisper** [comment]
I wasn't able to get the persistent code working. I'm thinking about starting a pull request for everything I've fixed so far. The only thing I'm worried about is whether duplication of keys should be allowed. It depends on what the keys are representing. If they are say representing prices, duplicates should be allowed. If the red-black tree is being used to represent a map, duplicates shouldn't be allowed. What is your take on this?

**Comment by hargettp** [comment]
Re persistent code: hah! Not surprised; its a bit of a mess, as it was my first attempt at a persistent data structure.  Oh, and first time implementing red-black trees, too. Thanks for trying!

Re duplicate keys: IIRC, I built it to work more like a map, which implies keys are not duplicated. Easier to balance if we assume no duplicates.

**Comment by malisper** [comment]
I made a compromise with respect to deduplicating the keys. By default the tree will deduplicate the keys, but there is an optional argument to allow duplicates. You should see a pull request come in shortly.
